// @azure/arm-compute/dist/commonjs/computeManagementClient.js
"use strict";
/*
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is regenerated.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ComputeManagementClient = void 0;
const tslib_1 = require("tslib");
const coreClient = tslib_1.__importStar(require("@azure/core-client"));
const coreRestPipeline = tslib_1.__importStar(require("@azure/core-rest-pipeline"));
const index_js_1 = require("./operations/index.js");
class ComputeManagementClient extends coreClient.ServiceClient {
    /**
     * Initializes a new instance of the ComputeManagementClient class.
     * @param credentials Subscription credentials which uniquely identify client subscription.
     * @param subscriptionId Subscription credentials which uniquely identify Microsoft Azure subscription.
     *                       The subscription ID forms part of the URI for every service call.
     * @param options The parameter options
     */
    constructor(credentials, subscriptionId, options) {
        var _a, _b, _c;
        if (credentials === undefined) {
            throw new Error("'credentials' cannot be null");
        }
        if (subscriptionId === undefined) {
            throw new Error("'subscriptionId' cannot be null");
        }
        // Initializing default values for options
        if (!options) {
            options = {};
        }
        const defaults = {
            requestContentType: "application/json; charset=utf-8",
            credential: credentials,
        };
        const packageDetails = `azsdk-js-arm-compute/22.4.0`;
        const userAgentPrefix = options.userAgentOptions && options.userAgentOptions.userAgentPrefix
            ? `${options.userAgentOptions.userAgentPrefix} ${packageDetails}`
            : `${packageDetails}`;
        const optionsWithDefaults = Object.assign(Object.assign(Object.assign({}, defaults), options), { userAgentOptions: {
                userAgentPrefix,
            }, endpoint: (_b = (_a = options.endpoint) !== null && _a !== void 0 ? _a : options.baseUri) !== null && _b !== void 0 ? _b : "https://management.azure.com" });
        super(optionsWithDefaults);
        let bearerTokenAuthenticationPolicyFound = false;
        if ((options === null || options === void 0 ? void 0 : options.pipeline) && options.pipeline.getOrderedPolicies().length > 0) {
            const pipelinePolicies = options.pipeline.getOrderedPolicies();
            bearerTokenAuthenticationPolicyFound = pipelinePolicies.some((pipelinePolicy) => pipelinePolicy.name ===
                coreRestPipeline.bearerTokenAuthenticationPolicyName);
        }
        if (!options ||
            !options.pipeline ||
            options.pipeline.getOrderedPolicies().length == 0 ||
            !bearerTokenAuthenticationPolicyFound) {
            this.pipeline.removePolicy({
                name: coreRestPipeline.bearerTokenAuthenticationPolicyName,
            });
            this.pipeline.addPolicy(coreRestPipeline.bearerTokenAuthenticationPolicy({
                credential: credentials,
                scopes: (_c = optionsWithDefaults.credentialScopes) !== null && _c !== void 0 ? _c : `${optionsWithDefaults.endpoint}/.default`,
                challengeCallbacks: {
                    authorizeRequestOnChallenge: coreClient.authorizeRequestOnClaimChallenge,
                },
            }));
        }
        // Parameter assignments
        this.subscriptionId = subscriptionId;
        // Assigning values to Constant parameters
        this.$host = options.$host || "https://management.azure.com";
        this.operations = new index_js_1.OperationsImpl(this);
        this.usageOperations = new index_js_1.UsageOperationsImpl(this);
        this.virtualMachineSizes = new index_js_1.VirtualMachineSizesImpl(this);
        this.virtualMachineScaleSets = new index_js_1.VirtualMachineScaleSetsImpl(this);
        this.virtualMachineScaleSetExtensions =
            new index_js_1.VirtualMachineScaleSetExtensionsImpl(this);
        this.virtualMachineScaleSetRollingUpgrades =
            new index_js_1.VirtualMachineScaleSetRollingUpgradesImpl(this);
        this.virtualMachineScaleSetVMExtensions =
            new index_js_1.VirtualMachineScaleSetVMExtensionsImpl(this);
        this.virtualMachineScaleSetVMs = new index_js_1.VirtualMachineScaleSetVMsImpl(this);
        this.virtualMachineExtensions = new index_js_1.VirtualMachineExtensionsImpl(this);
        this.virtualMachines = new index_js_1.VirtualMachinesImpl(this);
        this.virtualMachineImages = new index_js_1.VirtualMachineImagesImpl(this);
        this.virtualMachineImagesEdgeZone = new index_js_1.VirtualMachineImagesEdgeZoneImpl(this);
        this.virtualMachineExtensionImages = new index_js_1.VirtualMachineExtensionImagesImpl(this);
        this.availabilitySets = new index_js_1.AvailabilitySetsImpl(this);
        this.proximityPlacementGroups = new index_js_1.ProximityPlacementGroupsImpl(this);
        this.dedicatedHostGroups = new index_js_1.DedicatedHostGroupsImpl(this);
        this.dedicatedHosts = new index_js_1.DedicatedHostsImpl(this);
        this.sshPublicKeys = new index_js_1.SshPublicKeysImpl(this);
        this.images = new index_js_1.ImagesImpl(this);
        this.restorePointCollections = new index_js_1.RestorePointCollectionsImpl(this);
        this.restorePoints = new index_js_1.RestorePointsImpl(this);
        this.capacityReservationGroups = new index_js_1.CapacityReservationGroupsImpl(this);
        this.capacityReservations = new index_js_1.CapacityReservationsImpl(this);
        this.logAnalytics = new index_js_1.LogAnalyticsImpl(this);
        this.virtualMachineRunCommands = new index_js_1.VirtualMachineRunCommandsImpl(this);
        this.virtualMachineScaleSetVMRunCommands =
            new index_js_1.VirtualMachineScaleSetVMRunCommandsImpl(this);
        this.disks = new index_js_1.DisksImpl(this);
        this.diskAccesses = new index_js_1.DiskAccessesImpl(this);
        this.diskEncryptionSets = new index_js_1.DiskEncryptionSetsImpl(this);
        this.diskRestorePointOperations = new index_js_1.DiskRestorePointOperationsImpl(this);
        this.snapshots = new index_js_1.SnapshotsImpl(this);
        this.resourceSkus = new index_js_1.ResourceSkusImpl(this);
        this.galleries = new index_js_1.GalleriesImpl(this);
        this.galleryImages = new index_js_1.GalleryImagesImpl(this);
        this.galleryImageVersions = new index_js_1.GalleryImageVersionsImpl(this);
        this.galleryApplications = new index_js_1.GalleryApplicationsImpl(this);
        this.galleryApplicationVersions = new index_js_1.GalleryApplicationVersionsImpl(this);
        this.softDeletedResource = new index_js_1.SoftDeletedResourceImpl(this);
        this.gallerySharingProfile = new index_js_1.GallerySharingProfileImpl(this);
        this.galleryInVMAccessControlProfiles =
            new index_js_1.GalleryInVMAccessControlProfilesImpl(this);
        this.galleryInVMAccessControlProfileVersions =
            new index_js_1.GalleryInVMAccessControlProfileVersionsImpl(this);
        this.sharedGalleries = new index_js_1.SharedGalleriesImpl(this);
        this.sharedGalleryImages = new index_js_1.SharedGalleryImagesImpl(this);
        this.sharedGalleryImageVersions = new index_js_1.SharedGalleryImageVersionsImpl(this);
        this.communityGalleries = new index_js_1.CommunityGalleriesImpl(this);
        this.communityGalleryImages = new index_js_1.CommunityGalleryImagesImpl(this);
        this.communityGalleryImageVersions = new index_js_1.CommunityGalleryImageVersionsImpl(this);
        this.cloudServiceRoleInstances = new index_js_1.CloudServiceRoleInstancesImpl(this);
        this.cloudServiceRoles = new index_js_1.CloudServiceRolesImpl(this);
        this.cloudServices = new index_js_1.CloudServicesImpl(this);
        this.cloudServicesUpdateDomain = new index_js_1.CloudServicesUpdateDomainImpl(this);
        this.cloudServiceOperatingSystems = new index_js_1.CloudServiceOperatingSystemsImpl(this);
    }
}
exports.ComputeManagementClient = ComputeManagementClient;
