// @azure/arm-compute/dist/commonjs/operations/galleryInVMAccessControlProfiles.js
"use strict";
/*
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is regenerated.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.GalleryInVMAccessControlProfilesImpl = void 0;
const tslib_1 = require("tslib");
const pagingHelper_js_1 = require("../pagingHelper.js");
const coreClient = tslib_1.__importStar(require("@azure/core-client"));
const Mappers = tslib_1.__importStar(require("../models/mappers.js"));
const Parameters = tslib_1.__importStar(require("../models/parameters.js"));
const core_lro_1 = require("@azure/core-lro");
const lroImpl_js_1 = require("../lroImpl.js");
/// <reference lib="esnext.asynciterable" />
/** Class containing GalleryInVMAccessControlProfiles operations. */
class GalleryInVMAccessControlProfilesImpl {
    /**
     * Initialize a new instance of the class GalleryInVMAccessControlProfiles class.
     * @param client Reference to the service client
     */
    constructor(client) {
        this.client = client;
    }
    /**
     * List gallery inVMAccessControlProfiles in a gallery.
     * @param resourceGroupName The name of the resource group.
     * @param galleryName The name of the Shared Image Gallery from which the InVMAccessControlProfiles are
     *                    to be listed.
     * @param options The options parameters.
     */
    listByGallery(resourceGroupName, galleryName, options) {
        const iter = this.listByGalleryPagingAll(resourceGroupName, galleryName, options);
        return {
            next() {
                return iter.next();
            },
            [Symbol.asyncIterator]() {
                return this;
            },
            byPage: (settings) => {
                if (settings === null || settings === void 0 ? void 0 : settings.maxPageSize) {
                    throw new Error("maxPageSize is not supported by this operation.");
                }
                return this.listByGalleryPagingPage(resourceGroupName, galleryName, options, settings);
            },
        };
    }
    listByGalleryPagingPage(resourceGroupName, galleryName, options, settings) {
        return tslib_1.__asyncGenerator(this, arguments, function* listByGalleryPagingPage_1() {
            let result;
            let continuationToken = settings === null || settings === void 0 ? void 0 : settings.continuationToken;
            if (!continuationToken) {
                result = yield tslib_1.__await(this._listByGallery(resourceGroupName, galleryName, options));
                let page = result.value || [];
                continuationToken = result.nextLink;
                (0, pagingHelper_js_1.setContinuationToken)(page, continuationToken);
                yield yield tslib_1.__await(page);
            }
            while (continuationToken) {
                result = yield tslib_1.__await(this._listByGalleryNext(resourceGroupName, galleryName, continuationToken, options));
                continuationToken = result.nextLink;
                let page = result.value || [];
                (0, pagingHelper_js_1.setContinuationToken)(page, continuationToken);
                yield yield tslib_1.__await(page);
            }
        });
    }
    listByGalleryPagingAll(resourceGroupName, galleryName, options) {
        return tslib_1.__asyncGenerator(this, arguments, function* listByGalleryPagingAll_1() {
            var _a, e_1, _b, _c;
            try {
                for (var _d = true, _e = tslib_1.__asyncValues(this.listByGalleryPagingPage(resourceGroupName, galleryName, options)), _f; _f = yield tslib_1.__await(_e.next()), _a = _f.done, !_a; _d = true) {
                    _c = _f.value;
                    _d = false;
                    const page = _c;
                    yield tslib_1.__await(yield* tslib_1.__asyncDelegator(tslib_1.__asyncValues(page)));
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (!_d && !_a && (_b = _e.return)) yield tslib_1.__await(_b.call(_e));
                }
                finally { if (e_1) throw e_1.error; }
            }
        });
    }
    /**
     * Create or update a gallery inVMAccessControlProfile.
     * @param resourceGroupName The name of the resource group.
     * @param galleryName The name of the Shared Image Gallery in which the InVMAccessControlProfile is to
     *                    be created.
     * @param inVMAccessControlProfileName The name of the gallery inVMAccessControlProfile to be created
     *                                     or updated. The allowed characters are alphabets and numbers with dots, dashes, and periods allowed
     *                                     in the middle. The maximum length is 80 characters.
     * @param galleryInVMAccessControlProfile Parameters supplied to the create or update gallery
     *                                        inVMAccessControlProfile operation.
     * @param options The options parameters.
     */
    async beginCreateOrUpdate(resourceGroupName, galleryName, inVMAccessControlProfileName, galleryInVMAccessControlProfile, options) {
        const directSendOperation = async (args, spec) => {
            return this.client.sendOperationRequest(args, spec);
        };
        const sendOperationFn = async (args, spec) => {
            var _a;
            let currentRawResponse = undefined;
            const providedCallback = (_a = args.options) === null || _a === void 0 ? void 0 : _a.onResponse;
            const callback = (rawResponse, flatResponse) => {
                currentRawResponse = rawResponse;
                providedCallback === null || providedCallback === void 0 ? void 0 : providedCallback(rawResponse, flatResponse);
            };
            const updatedArgs = Object.assign(Object.assign({}, args), { options: Object.assign(Object.assign({}, args.options), { onResponse: callback }) });
            const flatResponse = await directSendOperation(updatedArgs, spec);
            return {
                flatResponse,
                rawResponse: {
                    statusCode: currentRawResponse.status,
                    body: currentRawResponse.parsedBody,
                    headers: currentRawResponse.headers.toJSON(),
                },
            };
        };
        const lro = (0, lroImpl_js_1.createLroSpec)({
            sendOperationFn,
            args: {
                resourceGroupName,
                galleryName,
                inVMAccessControlProfileName,
                galleryInVMAccessControlProfile,
                options,
            },
            spec: createOrUpdateOperationSpec,
        });
        const poller = await (0, core_lro_1.createHttpPoller)(lro, {
            restoreFrom: options === null || options === void 0 ? void 0 : options.resumeFrom,
            intervalInMs: options === null || options === void 0 ? void 0 : options.updateIntervalInMs,
        });
        await poller.poll();
        return poller;
    }
    /**
     * Create or update a gallery inVMAccessControlProfile.
     * @param resourceGroupName The name of the resource group.
     * @param galleryName The name of the Shared Image Gallery in which the InVMAccessControlProfile is to
     *                    be created.
     * @param inVMAccessControlProfileName The name of the gallery inVMAccessControlProfile to be created
     *                                     or updated. The allowed characters are alphabets and numbers with dots, dashes, and periods allowed
     *                                     in the middle. The maximum length is 80 characters.
     * @param galleryInVMAccessControlProfile Parameters supplied to the create or update gallery
     *                                        inVMAccessControlProfile operation.
     * @param options The options parameters.
     */
    async beginCreateOrUpdateAndWait(resourceGroupName, galleryName, inVMAccessControlProfileName, galleryInVMAccessControlProfile, options) {
        const poller = await this.beginCreateOrUpdate(resourceGroupName, galleryName, inVMAccessControlProfileName, galleryInVMAccessControlProfile, options);
        return poller.pollUntilDone();
    }
    /**
     * Update a gallery inVMAccessControlProfile.
     * @param resourceGroupName The name of the resource group.
     * @param galleryName The name of the Shared Image Gallery in which the InVMAccessControlProfile is to
     *                    be updated.
     * @param inVMAccessControlProfileName The name of the gallery inVMAccessControlProfile to be updated.
     *                                     The allowed characters are alphabets and numbers with dots, dashes, and periods allowed in the
     *                                     middle. The maximum length is 80 characters.
     * @param galleryInVMAccessControlProfile Parameters supplied to the update gallery
     *                                        inVMAccessControlProfile operation.
     * @param options The options parameters.
     */
    async beginUpdate(resourceGroupName, galleryName, inVMAccessControlProfileName, galleryInVMAccessControlProfile, options) {
        const directSendOperation = async (args, spec) => {
            return this.client.sendOperationRequest(args, spec);
        };
        const sendOperationFn = async (args, spec) => {
            var _a;
            let currentRawResponse = undefined;
            const providedCallback = (_a = args.options) === null || _a === void 0 ? void 0 : _a.onResponse;
            const callback = (rawResponse, flatResponse) => {
                currentRawResponse = rawResponse;
                providedCallback === null || providedCallback === void 0 ? void 0 : providedCallback(rawResponse, flatResponse);
            };
            const updatedArgs = Object.assign(Object.assign({}, args), { options: Object.assign(Object.assign({}, args.options), { onResponse: callback }) });
            const flatResponse = await directSendOperation(updatedArgs, spec);
            return {
                flatResponse,
                rawResponse: {
                    statusCode: currentRawResponse.status,
                    body: currentRawResponse.parsedBody,
                    headers: currentRawResponse.headers.toJSON(),
                },
            };
        };
        const lro = (0, lroImpl_js_1.createLroSpec)({
            sendOperationFn,
            args: {
                resourceGroupName,
                galleryName,
                inVMAccessControlProfileName,
                galleryInVMAccessControlProfile,
                options,
            },
            spec: updateOperationSpec,
        });
        const poller = await (0, core_lro_1.createHttpPoller)(lro, {
            restoreFrom: options === null || options === void 0 ? void 0 : options.resumeFrom,
            intervalInMs: options === null || options === void 0 ? void 0 : options.updateIntervalInMs,
        });
        await poller.poll();
        return poller;
    }
    /**
     * Update a gallery inVMAccessControlProfile.
     * @param resourceGroupName The name of the resource group.
     * @param galleryName The name of the Shared Image Gallery in which the InVMAccessControlProfile is to
     *                    be updated.
     * @param inVMAccessControlProfileName The name of the gallery inVMAccessControlProfile to be updated.
     *                                     The allowed characters are alphabets and numbers with dots, dashes, and periods allowed in the
     *                                     middle. The maximum length is 80 characters.
     * @param galleryInVMAccessControlProfile Parameters supplied to the update gallery
     *                                        inVMAccessControlProfile operation.
     * @param options The options parameters.
     */
    async beginUpdateAndWait(resourceGroupName, galleryName, inVMAccessControlProfileName, galleryInVMAccessControlProfile, options) {
        const poller = await this.beginUpdate(resourceGroupName, galleryName, inVMAccessControlProfileName, galleryInVMAccessControlProfile, options);
        return poller.pollUntilDone();
    }
    /**
     * Retrieves information about a gallery inVMAccessControlProfile.
     * @param resourceGroupName The name of the resource group.
     * @param galleryName The name of the Shared Image Gallery from which the InVMAccessControlProfiles are
     *                    to be retrieved.
     * @param inVMAccessControlProfileName The name of the gallery inVMAccessControlProfile to be
     *                                     retrieved.
     * @param options The options parameters.
     */
    get(resourceGroupName, galleryName, inVMAccessControlProfileName, options) {
        return this.client.sendOperationRequest({ resourceGroupName, galleryName, inVMAccessControlProfileName, options }, getOperationSpec);
    }
    /**
     * Delete a gallery inVMAccessControlProfile.
     * @param resourceGroupName The name of the resource group.
     * @param galleryName he name of the Shared Image Gallery in which the InVMAccessControlProfile
     *                    resides.
     * @param inVMAccessControlProfileName The name of the gallery inVMAccessControlProfile to be deleted.
     * @param options The options parameters.
     */
    async beginDelete(resourceGroupName, galleryName, inVMAccessControlProfileName, options) {
        const directSendOperation = async (args, spec) => {
            return this.client.sendOperationRequest(args, spec);
        };
        const sendOperationFn = async (args, spec) => {
            var _a;
            let currentRawResponse = undefined;
            const providedCallback = (_a = args.options) === null || _a === void 0 ? void 0 : _a.onResponse;
            const callback = (rawResponse, flatResponse) => {
                currentRawResponse = rawResponse;
                providedCallback === null || providedCallback === void 0 ? void 0 : providedCallback(rawResponse, flatResponse);
            };
            const updatedArgs = Object.assign(Object.assign({}, args), { options: Object.assign(Object.assign({}, args.options), { onResponse: callback }) });
            const flatResponse = await directSendOperation(updatedArgs, spec);
            return {
                flatResponse,
                rawResponse: {
                    statusCode: currentRawResponse.status,
                    body: currentRawResponse.parsedBody,
                    headers: currentRawResponse.headers.toJSON(),
                },
            };
        };
        const lro = (0, lroImpl_js_1.createLroSpec)({
            sendOperationFn,
            args: {
                resourceGroupName,
                galleryName,
                inVMAccessControlProfileName,
                options,
            },
            spec: deleteOperationSpec,
        });
        const poller = await (0, core_lro_1.createHttpPoller)(lro, {
            restoreFrom: options === null || options === void 0 ? void 0 : options.resumeFrom,
            intervalInMs: options === null || options === void 0 ? void 0 : options.updateIntervalInMs,
        });
        await poller.poll();
        return poller;
    }
    /**
     * Delete a gallery inVMAccessControlProfile.
     * @param resourceGroupName The name of the resource group.
     * @param galleryName he name of the Shared Image Gallery in which the InVMAccessControlProfile
     *                    resides.
     * @param inVMAccessControlProfileName The name of the gallery inVMAccessControlProfile to be deleted.
     * @param options The options parameters.
     */
    async beginDeleteAndWait(resourceGroupName, galleryName, inVMAccessControlProfileName, options) {
        const poller = await this.beginDelete(resourceGroupName, galleryName, inVMAccessControlProfileName, options);
        return poller.pollUntilDone();
    }
    /**
     * List gallery inVMAccessControlProfiles in a gallery.
     * @param resourceGroupName The name of the resource group.
     * @param galleryName The name of the Shared Image Gallery from which the InVMAccessControlProfiles are
     *                    to be listed.
     * @param options The options parameters.
     */
    _listByGallery(resourceGroupName, galleryName, options) {
        return this.client.sendOperationRequest({ resourceGroupName, galleryName, options }, listByGalleryOperationSpec);
    }
    /**
     * ListByGalleryNext
     * @param resourceGroupName The name of the resource group.
     * @param galleryName The name of the Shared Image Gallery from which the InVMAccessControlProfiles are
     *                    to be listed.
     * @param nextLink The nextLink from the previous successful call to the ListByGallery method.
     * @param options The options parameters.
     */
    _listByGalleryNext(resourceGroupName, galleryName, nextLink, options) {
        return this.client.sendOperationRequest({ resourceGroupName, galleryName, nextLink, options }, listByGalleryNextOperationSpec);
    }
}
exports.GalleryInVMAccessControlProfilesImpl = GalleryInVMAccessControlProfilesImpl;
// Operation Specifications
const serializer = coreClient.createSerializer(Mappers, /* isXml */ false);
const createOrUpdateOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Compute/galleries/{galleryName}/inVMAccessControlProfiles/{inVMAccessControlProfileName}",
    httpMethod: "PUT",
    responses: {
        200: {
            bodyMapper: Mappers.GalleryInVMAccessControlProfile,
        },
        201: {
            bodyMapper: Mappers.GalleryInVMAccessControlProfile,
        },
        202: {
            bodyMapper: Mappers.GalleryInVMAccessControlProfile,
        },
        204: {
            bodyMapper: Mappers.GalleryInVMAccessControlProfile,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    requestBody: Parameters.galleryInVMAccessControlProfile,
    queryParameters: [Parameters.apiVersion3],
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.resourceGroupName,
        Parameters.galleryName1,
        Parameters.inVMAccessControlProfileName,
    ],
    headerParameters: [Parameters.accept, Parameters.contentType],
    mediaType: "json",
    serializer,
};
const updateOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Compute/galleries/{galleryName}/inVMAccessControlProfiles/{inVMAccessControlProfileName}",
    httpMethod: "PATCH",
    responses: {
        200: {
            bodyMapper: Mappers.GalleryInVMAccessControlProfile,
        },
        201: {
            bodyMapper: Mappers.GalleryInVMAccessControlProfile,
        },
        202: {
            bodyMapper: Mappers.GalleryInVMAccessControlProfile,
        },
        204: {
            bodyMapper: Mappers.GalleryInVMAccessControlProfile,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    requestBody: Parameters.galleryInVMAccessControlProfile1,
    queryParameters: [Parameters.apiVersion3],
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.resourceGroupName,
        Parameters.galleryName,
        Parameters.inVMAccessControlProfileName1,
    ],
    headerParameters: [Parameters.accept, Parameters.contentType],
    mediaType: "json",
    serializer,
};
const getOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Compute/galleries/{galleryName}/inVMAccessControlProfiles/{inVMAccessControlProfileName}",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.GalleryInVMAccessControlProfile,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    queryParameters: [Parameters.apiVersion3],
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.resourceGroupName,
        Parameters.galleryName,
        Parameters.inVMAccessControlProfileName1,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const deleteOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Compute/galleries/{galleryName}/inVMAccessControlProfiles/{inVMAccessControlProfileName}",
    httpMethod: "DELETE",
    responses: {
        200: {
            headersMapper: Mappers.GalleryInVMAccessControlProfilesDeleteHeaders,
        },
        201: {
            headersMapper: Mappers.GalleryInVMAccessControlProfilesDeleteHeaders,
        },
        202: {
            headersMapper: Mappers.GalleryInVMAccessControlProfilesDeleteHeaders,
        },
        204: {
            headersMapper: Mappers.GalleryInVMAccessControlProfilesDeleteHeaders,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    queryParameters: [Parameters.apiVersion3],
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.resourceGroupName,
        Parameters.galleryName,
        Parameters.inVMAccessControlProfileName1,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const listByGalleryOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Compute/galleries/{galleryName}/inVMAccessControlProfiles",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.GalleryInVMAccessControlProfileList,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    queryParameters: [Parameters.apiVersion3],
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.resourceGroupName,
        Parameters.galleryName1,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const listByGalleryNextOperationSpec = {
    path: "{nextLink}",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.GalleryInVMAccessControlProfileList,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.nextLink,
        Parameters.resourceGroupName,
        Parameters.galleryName1,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
