// @azure/arm-compute/dist/commonjs/operations/galleryInVMAccessControlProfileVersions.js
"use strict";
/*
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is regenerated.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.GalleryInVMAccessControlProfileVersionsImpl = void 0;
const tslib_1 = require("tslib");
const pagingHelper_js_1 = require("../pagingHelper.js");
const coreClient = tslib_1.__importStar(require("@azure/core-client"));
const Mappers = tslib_1.__importStar(require("../models/mappers.js"));
const Parameters = tslib_1.__importStar(require("../models/parameters.js"));
const core_lro_1 = require("@azure/core-lro");
const lroImpl_js_1 = require("../lroImpl.js");
/// <reference lib="esnext.asynciterable" />
/** Class containing GalleryInVMAccessControlProfileVersions operations. */
class GalleryInVMAccessControlProfileVersionsImpl {
    /**
     * Initialize a new instance of the class GalleryInVMAccessControlProfileVersions class.
     * @param client Reference to the service client
     */
    constructor(client) {
        this.client = client;
    }
    /**
     * List gallery inVMAccessControlProfile versions in a gallery inVMAccessControlProfile
     * @param resourceGroupName The name of the resource group.
     * @param galleryName The name of the Shared Image Gallery in which the inVMAccessControlProfile
     *                    resides.
     * @param inVMAccessControlProfileName The name of the gallery inVMAccessControlProfile from which the
     *                                     inVMAccessControlProfile versions are to be listed.
     * @param options The options parameters.
     */
    listByGalleryInVMAccessControlProfile(resourceGroupName, galleryName, inVMAccessControlProfileName, options) {
        const iter = this.listByGalleryInVMAccessControlProfilePagingAll(resourceGroupName, galleryName, inVMAccessControlProfileName, options);
        return {
            next() {
                return iter.next();
            },
            [Symbol.asyncIterator]() {
                return this;
            },
            byPage: (settings) => {
                if (settings === null || settings === void 0 ? void 0 : settings.maxPageSize) {
                    throw new Error("maxPageSize is not supported by this operation.");
                }
                return this.listByGalleryInVMAccessControlProfilePagingPage(resourceGroupName, galleryName, inVMAccessControlProfileName, options, settings);
            },
        };
    }
    listByGalleryInVMAccessControlProfilePagingPage(resourceGroupName, galleryName, inVMAccessControlProfileName, options, settings) {
        return tslib_1.__asyncGenerator(this, arguments, function* listByGalleryInVMAccessControlProfilePagingPage_1() {
            let result;
            let continuationToken = settings === null || settings === void 0 ? void 0 : settings.continuationToken;
            if (!continuationToken) {
                result = yield tslib_1.__await(this._listByGalleryInVMAccessControlProfile(resourceGroupName, galleryName, inVMAccessControlProfileName, options));
                let page = result.value || [];
                continuationToken = result.nextLink;
                (0, pagingHelper_js_1.setContinuationToken)(page, continuationToken);
                yield yield tslib_1.__await(page);
            }
            while (continuationToken) {
                result = yield tslib_1.__await(this._listByGalleryInVMAccessControlProfileNext(resourceGroupName, galleryName, inVMAccessControlProfileName, continuationToken, options));
                continuationToken = result.nextLink;
                let page = result.value || [];
                (0, pagingHelper_js_1.setContinuationToken)(page, continuationToken);
                yield yield tslib_1.__await(page);
            }
        });
    }
    listByGalleryInVMAccessControlProfilePagingAll(resourceGroupName, galleryName, inVMAccessControlProfileName, options) {
        return tslib_1.__asyncGenerator(this, arguments, function* listByGalleryInVMAccessControlProfilePagingAll_1() {
            var _a, e_1, _b, _c;
            try {
                for (var _d = true, _e = tslib_1.__asyncValues(this.listByGalleryInVMAccessControlProfilePagingPage(resourceGroupName, galleryName, inVMAccessControlProfileName, options)), _f; _f = yield tslib_1.__await(_e.next()), _a = _f.done, !_a; _d = true) {
                    _c = _f.value;
                    _d = false;
                    const page = _c;
                    yield tslib_1.__await(yield* tslib_1.__asyncDelegator(tslib_1.__asyncValues(page)));
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (!_d && !_a && (_b = _e.return)) yield tslib_1.__await(_b.call(_e));
                }
                finally { if (e_1) throw e_1.error; }
            }
        });
    }
    /**
     * Create or update a gallery inVMAccessControlProfile version.
     * @param resourceGroupName The name of the resource group.
     * @param galleryName The name of the Shared Image Gallery in which the inVMAccessControlProfile
     *                    resides.
     * @param inVMAccessControlProfileName The name of the gallery inVMAccessControlProfile in which the
     *                                     inVMAccessControlProfile version is to be created.
     * @param inVMAccessControlProfileVersionName The name of the gallery inVMAccessControlProfile version
     *                                            to be created. Needs to follow semantic version name pattern: The allowed characters are digit and
     *                                            period. Digits must be within the range of a 32-bit integer. Format:
     *                                            <MajorVersion>.<MinorVersion>.<Patch>
     * @param galleryInVMAccessControlProfileVersion Parameters supplied to the create or update gallery
     *                                               inVMAccessControlProfile version operation.
     * @param options The options parameters.
     */
    async beginCreateOrUpdate(resourceGroupName, galleryName, inVMAccessControlProfileName, inVMAccessControlProfileVersionName, galleryInVMAccessControlProfileVersion, options) {
        const directSendOperation = async (args, spec) => {
            return this.client.sendOperationRequest(args, spec);
        };
        const sendOperationFn = async (args, spec) => {
            var _a;
            let currentRawResponse = undefined;
            const providedCallback = (_a = args.options) === null || _a === void 0 ? void 0 : _a.onResponse;
            const callback = (rawResponse, flatResponse) => {
                currentRawResponse = rawResponse;
                providedCallback === null || providedCallback === void 0 ? void 0 : providedCallback(rawResponse, flatResponse);
            };
            const updatedArgs = Object.assign(Object.assign({}, args), { options: Object.assign(Object.assign({}, args.options), { onResponse: callback }) });
            const flatResponse = await directSendOperation(updatedArgs, spec);
            return {
                flatResponse,
                rawResponse: {
                    statusCode: currentRawResponse.status,
                    body: currentRawResponse.parsedBody,
                    headers: currentRawResponse.headers.toJSON(),
                },
            };
        };
        const lro = (0, lroImpl_js_1.createLroSpec)({
            sendOperationFn,
            args: {
                resourceGroupName,
                galleryName,
                inVMAccessControlProfileName,
                inVMAccessControlProfileVersionName,
                galleryInVMAccessControlProfileVersion,
                options,
            },
            spec: createOrUpdateOperationSpec,
        });
        const poller = await (0, core_lro_1.createHttpPoller)(lro, {
            restoreFrom: options === null || options === void 0 ? void 0 : options.resumeFrom,
            intervalInMs: options === null || options === void 0 ? void 0 : options.updateIntervalInMs,
        });
        await poller.poll();
        return poller;
    }
    /**
     * Create or update a gallery inVMAccessControlProfile version.
     * @param resourceGroupName The name of the resource group.
     * @param galleryName The name of the Shared Image Gallery in which the inVMAccessControlProfile
     *                    resides.
     * @param inVMAccessControlProfileName The name of the gallery inVMAccessControlProfile in which the
     *                                     inVMAccessControlProfile version is to be created.
     * @param inVMAccessControlProfileVersionName The name of the gallery inVMAccessControlProfile version
     *                                            to be created. Needs to follow semantic version name pattern: The allowed characters are digit and
     *                                            period. Digits must be within the range of a 32-bit integer. Format:
     *                                            <MajorVersion>.<MinorVersion>.<Patch>
     * @param galleryInVMAccessControlProfileVersion Parameters supplied to the create or update gallery
     *                                               inVMAccessControlProfile version operation.
     * @param options The options parameters.
     */
    async beginCreateOrUpdateAndWait(resourceGroupName, galleryName, inVMAccessControlProfileName, inVMAccessControlProfileVersionName, galleryInVMAccessControlProfileVersion, options) {
        const poller = await this.beginCreateOrUpdate(resourceGroupName, galleryName, inVMAccessControlProfileName, inVMAccessControlProfileVersionName, galleryInVMAccessControlProfileVersion, options);
        return poller.pollUntilDone();
    }
    /**
     * Update a gallery inVMAccessControlProfile version.
     * @param resourceGroupName The name of the resource group.
     * @param galleryName The name of the Shared Image Gallery in which the inVMAccessControlProfile
     *                    resides.
     * @param inVMAccessControlProfileName The name of the gallery inVMAccessControlProfile in which the
     *                                     inVMAccessControlProfile version is to be updated.
     * @param inVMAccessControlProfileVersionName The name of the gallery inVMAccessControlProfile version
     *                                            to be updated. Needs to follow semantic version name pattern: The allowed characters are digit and
     *                                            period. Digits must be within the range of a 32-bit integer. Format:
     *                                            <MajorVersion>.<MinorVersion>.<Patch>
     * @param galleryInVMAccessControlProfileVersion Parameters supplied to the update gallery
     *                                               inVMAccessControlProfile version operation.
     * @param options The options parameters.
     */
    async beginUpdate(resourceGroupName, galleryName, inVMAccessControlProfileName, inVMAccessControlProfileVersionName, galleryInVMAccessControlProfileVersion, options) {
        const directSendOperation = async (args, spec) => {
            return this.client.sendOperationRequest(args, spec);
        };
        const sendOperationFn = async (args, spec) => {
            var _a;
            let currentRawResponse = undefined;
            const providedCallback = (_a = args.options) === null || _a === void 0 ? void 0 : _a.onResponse;
            const callback = (rawResponse, flatResponse) => {
                currentRawResponse = rawResponse;
                providedCallback === null || providedCallback === void 0 ? void 0 : providedCallback(rawResponse, flatResponse);
            };
            const updatedArgs = Object.assign(Object.assign({}, args), { options: Object.assign(Object.assign({}, args.options), { onResponse: callback }) });
            const flatResponse = await directSendOperation(updatedArgs, spec);
            return {
                flatResponse,
                rawResponse: {
                    statusCode: currentRawResponse.status,
                    body: currentRawResponse.parsedBody,
                    headers: currentRawResponse.headers.toJSON(),
                },
            };
        };
        const lro = (0, lroImpl_js_1.createLroSpec)({
            sendOperationFn,
            args: {
                resourceGroupName,
                galleryName,
                inVMAccessControlProfileName,
                inVMAccessControlProfileVersionName,
                galleryInVMAccessControlProfileVersion,
                options,
            },
            spec: updateOperationSpec,
        });
        const poller = await (0, core_lro_1.createHttpPoller)(lro, {
            restoreFrom: options === null || options === void 0 ? void 0 : options.resumeFrom,
            intervalInMs: options === null || options === void 0 ? void 0 : options.updateIntervalInMs,
        });
        await poller.poll();
        return poller;
    }
    /**
     * Update a gallery inVMAccessControlProfile version.
     * @param resourceGroupName The name of the resource group.
     * @param galleryName The name of the Shared Image Gallery in which the inVMAccessControlProfile
     *                    resides.
     * @param inVMAccessControlProfileName The name of the gallery inVMAccessControlProfile in which the
     *                                     inVMAccessControlProfile version is to be updated.
     * @param inVMAccessControlProfileVersionName The name of the gallery inVMAccessControlProfile version
     *                                            to be updated. Needs to follow semantic version name pattern: The allowed characters are digit and
     *                                            period. Digits must be within the range of a 32-bit integer. Format:
     *                                            <MajorVersion>.<MinorVersion>.<Patch>
     * @param galleryInVMAccessControlProfileVersion Parameters supplied to the update gallery
     *                                               inVMAccessControlProfile version operation.
     * @param options The options parameters.
     */
    async beginUpdateAndWait(resourceGroupName, galleryName, inVMAccessControlProfileName, inVMAccessControlProfileVersionName, galleryInVMAccessControlProfileVersion, options) {
        const poller = await this.beginUpdate(resourceGroupName, galleryName, inVMAccessControlProfileName, inVMAccessControlProfileVersionName, galleryInVMAccessControlProfileVersion, options);
        return poller.pollUntilDone();
    }
    /**
     * Retrieves information about a gallery inVMAccessControlProfile version.
     * @param resourceGroupName The name of the resource group.
     * @param galleryName The name of the Shared Image Gallery in which the inVMAccessControlProfile
     *                    resides.
     * @param inVMAccessControlProfileName The name of the gallery inVMAccessControlProfile in which the
     *                                     inVMAccessControlProfile version resides.
     * @param inVMAccessControlProfileVersionName The name of the gallery inVMAccessControlProfile version
     *                                            to be retrieved.
     * @param options The options parameters.
     */
    get(resourceGroupName, galleryName, inVMAccessControlProfileName, inVMAccessControlProfileVersionName, options) {
        return this.client.sendOperationRequest({
            resourceGroupName,
            galleryName,
            inVMAccessControlProfileName,
            inVMAccessControlProfileVersionName,
            options,
        }, getOperationSpec);
    }
    /**
     * Delete a gallery inVMAccessControlProfile version.
     * @param resourceGroupName The name of the resource group.
     * @param galleryName The name of the Shared Image Gallery in which the inVMAccessControlProfile
     *                    resides.
     * @param inVMAccessControlProfileName The name of the gallery inVMAccessControlProfile in which the
     *                                     inVMAccessControlProfile version resides.
     * @param inVMAccessControlProfileVersionName The name of the gallery inVMAccessControlProfile version
     *                                            to be deleted.
     * @param options The options parameters.
     */
    async beginDelete(resourceGroupName, galleryName, inVMAccessControlProfileName, inVMAccessControlProfileVersionName, options) {
        const directSendOperation = async (args, spec) => {
            return this.client.sendOperationRequest(args, spec);
        };
        const sendOperationFn = async (args, spec) => {
            var _a;
            let currentRawResponse = undefined;
            const providedCallback = (_a = args.options) === null || _a === void 0 ? void 0 : _a.onResponse;
            const callback = (rawResponse, flatResponse) => {
                currentRawResponse = rawResponse;
                providedCallback === null || providedCallback === void 0 ? void 0 : providedCallback(rawResponse, flatResponse);
            };
            const updatedArgs = Object.assign(Object.assign({}, args), { options: Object.assign(Object.assign({}, args.options), { onResponse: callback }) });
            const flatResponse = await directSendOperation(updatedArgs, spec);
            return {
                flatResponse,
                rawResponse: {
                    statusCode: currentRawResponse.status,
                    body: currentRawResponse.parsedBody,
                    headers: currentRawResponse.headers.toJSON(),
                },
            };
        };
        const lro = (0, lroImpl_js_1.createLroSpec)({
            sendOperationFn,
            args: {
                resourceGroupName,
                galleryName,
                inVMAccessControlProfileName,
                inVMAccessControlProfileVersionName,
                options,
            },
            spec: deleteOperationSpec,
        });
        const poller = await (0, core_lro_1.createHttpPoller)(lro, {
            restoreFrom: options === null || options === void 0 ? void 0 : options.resumeFrom,
            intervalInMs: options === null || options === void 0 ? void 0 : options.updateIntervalInMs,
        });
        await poller.poll();
        return poller;
    }
    /**
     * Delete a gallery inVMAccessControlProfile version.
     * @param resourceGroupName The name of the resource group.
     * @param galleryName The name of the Shared Image Gallery in which the inVMAccessControlProfile
     *                    resides.
     * @param inVMAccessControlProfileName The name of the gallery inVMAccessControlProfile in which the
     *                                     inVMAccessControlProfile version resides.
     * @param inVMAccessControlProfileVersionName The name of the gallery inVMAccessControlProfile version
     *                                            to be deleted.
     * @param options The options parameters.
     */
    async beginDeleteAndWait(resourceGroupName, galleryName, inVMAccessControlProfileName, inVMAccessControlProfileVersionName, options) {
        const poller = await this.beginDelete(resourceGroupName, galleryName, inVMAccessControlProfileName, inVMAccessControlProfileVersionName, options);
        return poller.pollUntilDone();
    }
    /**
     * List gallery inVMAccessControlProfile versions in a gallery inVMAccessControlProfile
     * @param resourceGroupName The name of the resource group.
     * @param galleryName The name of the Shared Image Gallery in which the inVMAccessControlProfile
     *                    resides.
     * @param inVMAccessControlProfileName The name of the gallery inVMAccessControlProfile from which the
     *                                     inVMAccessControlProfile versions are to be listed.
     * @param options The options parameters.
     */
    _listByGalleryInVMAccessControlProfile(resourceGroupName, galleryName, inVMAccessControlProfileName, options) {
        return this.client.sendOperationRequest({ resourceGroupName, galleryName, inVMAccessControlProfileName, options }, listByGalleryInVMAccessControlProfileOperationSpec);
    }
    /**
     * ListByGalleryInVMAccessControlProfileNext
     * @param resourceGroupName The name of the resource group.
     * @param galleryName The name of the Shared Image Gallery in which the inVMAccessControlProfile
     *                    resides.
     * @param inVMAccessControlProfileName The name of the gallery inVMAccessControlProfile from which the
     *                                     inVMAccessControlProfile versions are to be listed.
     * @param nextLink The nextLink from the previous successful call to the
     *                 ListByGalleryInVMAccessControlProfile method.
     * @param options The options parameters.
     */
    _listByGalleryInVMAccessControlProfileNext(resourceGroupName, galleryName, inVMAccessControlProfileName, nextLink, options) {
        return this.client.sendOperationRequest({
            resourceGroupName,
            galleryName,
            inVMAccessControlProfileName,
            nextLink,
            options,
        }, listByGalleryInVMAccessControlProfileNextOperationSpec);
    }
}
exports.GalleryInVMAccessControlProfileVersionsImpl = GalleryInVMAccessControlProfileVersionsImpl;
// Operation Specifications
const serializer = coreClient.createSerializer(Mappers, /* isXml */ false);
const createOrUpdateOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Compute/galleries/{galleryName}/inVMAccessControlProfiles/{inVMAccessControlProfileName}/versions/{inVMAccessControlProfileVersionName}",
    httpMethod: "PUT",
    responses: {
        200: {
            bodyMapper: Mappers.GalleryInVMAccessControlProfileVersion,
        },
        201: {
            bodyMapper: Mappers.GalleryInVMAccessControlProfileVersion,
        },
        202: {
            bodyMapper: Mappers.GalleryInVMAccessControlProfileVersion,
        },
        204: {
            bodyMapper: Mappers.GalleryInVMAccessControlProfileVersion,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    requestBody: Parameters.galleryInVMAccessControlProfileVersion,
    queryParameters: [Parameters.apiVersion3],
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.resourceGroupName,
        Parameters.galleryName1,
        Parameters.inVMAccessControlProfileName,
        Parameters.inVMAccessControlProfileVersionName,
    ],
    headerParameters: [Parameters.accept, Parameters.contentType],
    mediaType: "json",
    serializer,
};
const updateOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Compute/galleries/{galleryName}/inVMAccessControlProfiles/{inVMAccessControlProfileName}/versions/{inVMAccessControlProfileVersionName}",
    httpMethod: "PATCH",
    responses: {
        200: {
            bodyMapper: Mappers.GalleryInVMAccessControlProfileVersion,
        },
        201: {
            bodyMapper: Mappers.GalleryInVMAccessControlProfileVersion,
        },
        202: {
            bodyMapper: Mappers.GalleryInVMAccessControlProfileVersion,
        },
        204: {
            bodyMapper: Mappers.GalleryInVMAccessControlProfileVersion,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    requestBody: Parameters.galleryInVMAccessControlProfileVersion1,
    queryParameters: [Parameters.apiVersion3],
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.resourceGroupName,
        Parameters.galleryName,
        Parameters.inVMAccessControlProfileName1,
        Parameters.inVMAccessControlProfileVersionName1,
    ],
    headerParameters: [Parameters.accept, Parameters.contentType],
    mediaType: "json",
    serializer,
};
const getOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Compute/galleries/{galleryName}/inVMAccessControlProfiles/{inVMAccessControlProfileName}/versions/{inVMAccessControlProfileVersionName}",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.GalleryInVMAccessControlProfileVersion,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    queryParameters: [Parameters.apiVersion3],
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.resourceGroupName,
        Parameters.galleryName,
        Parameters.inVMAccessControlProfileName1,
        Parameters.inVMAccessControlProfileVersionName1,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const deleteOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Compute/galleries/{galleryName}/inVMAccessControlProfiles/{inVMAccessControlProfileName}/versions/{inVMAccessControlProfileVersionName}",
    httpMethod: "DELETE",
    responses: {
        200: {
            headersMapper: Mappers.GalleryInVMAccessControlProfileVersionsDeleteHeaders,
        },
        201: {
            headersMapper: Mappers.GalleryInVMAccessControlProfileVersionsDeleteHeaders,
        },
        202: {
            headersMapper: Mappers.GalleryInVMAccessControlProfileVersionsDeleteHeaders,
        },
        204: {
            headersMapper: Mappers.GalleryInVMAccessControlProfileVersionsDeleteHeaders,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    queryParameters: [Parameters.apiVersion3],
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.resourceGroupName,
        Parameters.galleryName,
        Parameters.inVMAccessControlProfileName1,
        Parameters.inVMAccessControlProfileVersionName1,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const listByGalleryInVMAccessControlProfileOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Compute/galleries/{galleryName}/inVMAccessControlProfiles/{inVMAccessControlProfileName}/versions",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.GalleryInVMAccessControlProfileVersionList,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    queryParameters: [Parameters.apiVersion3],
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.resourceGroupName,
        Parameters.galleryName1,
        Parameters.inVMAccessControlProfileName2,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const listByGalleryInVMAccessControlProfileNextOperationSpec = {
    path: "{nextLink}",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.GalleryInVMAccessControlProfileVersionList,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.nextLink,
        Parameters.resourceGroupName,
        Parameters.galleryName1,
        Parameters.inVMAccessControlProfileName2,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
