// @azure/arm-compute/dist/commonjs/operations/softDeletedResource.js
"use strict";
/*
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is regenerated.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SoftDeletedResourceImpl = void 0;
const tslib_1 = require("tslib");
const pagingHelper_js_1 = require("../pagingHelper.js");
const coreClient = tslib_1.__importStar(require("@azure/core-client"));
const Mappers = tslib_1.__importStar(require("../models/mappers.js"));
const Parameters = tslib_1.__importStar(require("../models/parameters.js"));
/// <reference lib="esnext.asynciterable" />
/** Class containing SoftDeletedResource operations. */
class SoftDeletedResourceImpl {
    /**
     * Initialize a new instance of the class SoftDeletedResource class.
     * @param client Reference to the service client
     */
    constructor(client) {
        this.client = client;
    }
    /**
     * List soft-deleted resources of an artifact in the gallery, such as soft-deleted gallery image
     * version of an image.
     * @param resourceGroupName The name of the resource group.
     * @param galleryName The name of the Gallery in which the soft-deleted resources resides.
     * @param artifactType The type of the artifact to be listed, such as gallery image version.
     * @param artifactName The artifact name to be listed. If artifact type is Images, then the artifact
     *                     name should be the gallery image name.
     * @param options The options parameters.
     */
    listByArtifactName(resourceGroupName, galleryName, artifactType, artifactName, options) {
        const iter = this.listByArtifactNamePagingAll(resourceGroupName, galleryName, artifactType, artifactName, options);
        return {
            next() {
                return iter.next();
            },
            [Symbol.asyncIterator]() {
                return this;
            },
            byPage: (settings) => {
                if (settings === null || settings === void 0 ? void 0 : settings.maxPageSize) {
                    throw new Error("maxPageSize is not supported by this operation.");
                }
                return this.listByArtifactNamePagingPage(resourceGroupName, galleryName, artifactType, artifactName, options, settings);
            },
        };
    }
    listByArtifactNamePagingPage(resourceGroupName, galleryName, artifactType, artifactName, options, settings) {
        return tslib_1.__asyncGenerator(this, arguments, function* listByArtifactNamePagingPage_1() {
            let result;
            let continuationToken = settings === null || settings === void 0 ? void 0 : settings.continuationToken;
            if (!continuationToken) {
                result = yield tslib_1.__await(this._listByArtifactName(resourceGroupName, galleryName, artifactType, artifactName, options));
                let page = result.value || [];
                continuationToken = result.nextLink;
                (0, pagingHelper_js_1.setContinuationToken)(page, continuationToken);
                yield yield tslib_1.__await(page);
            }
            while (continuationToken) {
                result = yield tslib_1.__await(this._listByArtifactNameNext(resourceGroupName, galleryName, artifactType, artifactName, continuationToken, options));
                continuationToken = result.nextLink;
                let page = result.value || [];
                (0, pagingHelper_js_1.setContinuationToken)(page, continuationToken);
                yield yield tslib_1.__await(page);
            }
        });
    }
    listByArtifactNamePagingAll(resourceGroupName, galleryName, artifactType, artifactName, options) {
        return tslib_1.__asyncGenerator(this, arguments, function* listByArtifactNamePagingAll_1() {
            var _a, e_1, _b, _c;
            try {
                for (var _d = true, _e = tslib_1.__asyncValues(this.listByArtifactNamePagingPage(resourceGroupName, galleryName, artifactType, artifactName, options)), _f; _f = yield tslib_1.__await(_e.next()), _a = _f.done, !_a; _d = true) {
                    _c = _f.value;
                    _d = false;
                    const page = _c;
                    yield tslib_1.__await(yield* tslib_1.__asyncDelegator(tslib_1.__asyncValues(page)));
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (!_d && !_a && (_b = _e.return)) yield tslib_1.__await(_b.call(_e));
                }
                finally { if (e_1) throw e_1.error; }
            }
        });
    }
    /**
     * List soft-deleted resources of an artifact in the gallery, such as soft-deleted gallery image
     * version of an image.
     * @param resourceGroupName The name of the resource group.
     * @param galleryName The name of the Gallery in which the soft-deleted resources resides.
     * @param artifactType The type of the artifact to be listed, such as gallery image version.
     * @param artifactName The artifact name to be listed. If artifact type is Images, then the artifact
     *                     name should be the gallery image name.
     * @param options The options parameters.
     */
    _listByArtifactName(resourceGroupName, galleryName, artifactType, artifactName, options) {
        return this.client.sendOperationRequest({ resourceGroupName, galleryName, artifactType, artifactName, options }, listByArtifactNameOperationSpec);
    }
    /**
     * ListByArtifactNameNext
     * @param resourceGroupName The name of the resource group.
     * @param galleryName The name of the Gallery in which the soft-deleted resources resides.
     * @param artifactType The type of the artifact to be listed, such as gallery image version.
     * @param artifactName The artifact name to be listed. If artifact type is Images, then the artifact
     *                     name should be the gallery image name.
     * @param nextLink The nextLink from the previous successful call to the ListByArtifactName method.
     * @param options The options parameters.
     */
    _listByArtifactNameNext(resourceGroupName, galleryName, artifactType, artifactName, nextLink, options) {
        return this.client.sendOperationRequest({
            resourceGroupName,
            galleryName,
            artifactType,
            artifactName,
            nextLink,
            options,
        }, listByArtifactNameNextOperationSpec);
    }
}
exports.SoftDeletedResourceImpl = SoftDeletedResourceImpl;
// Operation Specifications
const serializer = coreClient.createSerializer(Mappers, /* isXml */ false);
const listByArtifactNameOperationSpec = {
    path: "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Compute/galleries/{galleryName}/softDeletedArtifactTypes/{artifactType}/artifacts/{artifactName}/versions",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.GallerySoftDeletedResourceList,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    queryParameters: [Parameters.apiVersion3],
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.resourceGroupName,
        Parameters.galleryName1,
        Parameters.artifactType,
        Parameters.artifactName,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const listByArtifactNameNextOperationSpec = {
    path: "{nextLink}",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.GallerySoftDeletedResourceList,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.nextLink,
        Parameters.resourceGroupName,
        Parameters.galleryName1,
        Parameters.artifactType,
        Parameters.artifactName,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
