// @azure/arm-compute/dist/commonjs/operations/virtualMachineImages.js
"use strict";
/*
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is regenerated.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.VirtualMachineImagesImpl = void 0;
const tslib_1 = require("tslib");
const pagingHelper_js_1 = require("../pagingHelper.js");
const coreClient = tslib_1.__importStar(require("@azure/core-client"));
const Mappers = tslib_1.__importStar(require("../models/mappers.js"));
const Parameters = tslib_1.__importStar(require("../models/parameters.js"));
/// <reference lib="esnext.asynciterable" />
/** Class containing VirtualMachineImages operations. */
class VirtualMachineImagesImpl {
    /**
     * Initialize a new instance of the class VirtualMachineImages class.
     * @param client Reference to the service client
     */
    constructor(client) {
        this.client = client;
    }
    /**
     * @param location The name of a supported Azure region.
     * @param publisherName A valid image publisher.
     * @param offer A valid image publisher offer.
     * @param skus A valid image SKU.
     * @param expand The expand expression to apply on the operation.
     * @param options The options parameters.
     */
    listWithProperties(location, publisherName, offer, skus, expand, options) {
        const iter = this.listWithPropertiesPagingAll(location, publisherName, offer, skus, expand, options);
        return {
            next() {
                return iter.next();
            },
            [Symbol.asyncIterator]() {
                return this;
            },
            byPage: (settings) => {
                if (settings === null || settings === void 0 ? void 0 : settings.maxPageSize) {
                    throw new Error("maxPageSize is not supported by this operation.");
                }
                return this.listWithPropertiesPagingPage(location, publisherName, offer, skus, expand, options, settings);
            },
        };
    }
    listWithPropertiesPagingPage(location, publisherName, offer, skus, expand, options, settings) {
        return tslib_1.__asyncGenerator(this, arguments, function* listWithPropertiesPagingPage_1() {
            let result;
            let continuationToken = settings === null || settings === void 0 ? void 0 : settings.continuationToken;
            if (!continuationToken) {
                result = yield tslib_1.__await(this._listWithProperties(location, publisherName, offer, skus, expand, options));
                let page = result.value || [];
                continuationToken = result.nextLink;
                (0, pagingHelper_js_1.setContinuationToken)(page, continuationToken);
                yield yield tslib_1.__await(page);
            }
            while (continuationToken) {
                result = yield tslib_1.__await(this._listWithPropertiesNext(location, publisherName, offer, skus, continuationToken, options));
                continuationToken = result.nextLink;
                let page = result.value || [];
                (0, pagingHelper_js_1.setContinuationToken)(page, continuationToken);
                yield yield tslib_1.__await(page);
            }
        });
    }
    listWithPropertiesPagingAll(location, publisherName, offer, skus, expand, options) {
        return tslib_1.__asyncGenerator(this, arguments, function* listWithPropertiesPagingAll_1() {
            var _a, e_1, _b, _c;
            try {
                for (var _d = true, _e = tslib_1.__asyncValues(this.listWithPropertiesPagingPage(location, publisherName, offer, skus, expand, options)), _f; _f = yield tslib_1.__await(_e.next()), _a = _f.done, !_a; _d = true) {
                    _c = _f.value;
                    _d = false;
                    const page = _c;
                    yield tslib_1.__await(yield* tslib_1.__asyncDelegator(tslib_1.__asyncValues(page)));
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (!_d && !_a && (_b = _e.return)) yield tslib_1.__await(_b.call(_e));
                }
                finally { if (e_1) throw e_1.error; }
            }
        });
    }
    /**
     * Gets a virtual machine image.
     * @param location The name of a supported Azure region.
     * @param publisherName A valid image publisher.
     * @param offer A valid image publisher offer.
     * @param skus A valid image SKU.
     * @param version A valid image SKU version.
     * @param options The options parameters.
     */
    get(location, publisherName, offer, skus, version, options) {
        return this.client.sendOperationRequest({ location, publisherName, offer, skus, version, options }, getOperationSpec);
    }
    /**
     * Gets a list of all virtual machine image versions for the specified location, publisher, offer, and
     * SKU.
     * @param location The name of a supported Azure region.
     * @param publisherName A valid image publisher.
     * @param offer A valid image publisher offer.
     * @param skus A valid image SKU.
     * @param options The options parameters.
     */
    list(location, publisherName, offer, skus, options) {
        return this.client.sendOperationRequest({ location, publisherName, offer, skus, options }, listOperationSpec);
    }
    /**
     * Gets a list of virtual machine image offers for the specified location and publisher.
     * @param location The name of a supported Azure region.
     * @param publisherName A valid image publisher.
     * @param options The options parameters.
     */
    listOffers(location, publisherName, options) {
        return this.client.sendOperationRequest({ location, publisherName, options }, listOffersOperationSpec);
    }
    /**
     * Gets a list of virtual machine image publishers for the specified Azure location.
     * @param location The name of a supported Azure region.
     * @param options The options parameters.
     */
    listPublishers(location, options) {
        return this.client.sendOperationRequest({ location, options }, listPublishersOperationSpec);
    }
    /**
     * Gets a list of virtual machine image SKUs for the specified location, publisher, and offer.
     * @param location The name of a supported Azure region.
     * @param publisherName A valid image publisher.
     * @param offer A valid image publisher offer.
     * @param options The options parameters.
     */
    listSkus(location, publisherName, offer, options) {
        return this.client.sendOperationRequest({ location, publisherName, offer, options }, listSkusOperationSpec);
    }
    /**
     * Gets a list of all virtual machine image versions for the specified edge zone
     * @param location The name of a supported Azure region.
     * @param edgeZone The name of the edge zone.
     * @param options The options parameters.
     */
    listByEdgeZone(location, edgeZone, options) {
        return this.client.sendOperationRequest({ location, edgeZone, options }, listByEdgeZoneOperationSpec);
    }
    /**
     * @param location The name of a supported Azure region.
     * @param publisherName A valid image publisher.
     * @param offer A valid image publisher offer.
     * @param skus A valid image SKU.
     * @param expand The expand expression to apply on the operation.
     * @param options The options parameters.
     */
    _listWithProperties(location, publisherName, offer, skus, expand, options) {
        return this.client.sendOperationRequest({ location, publisherName, offer, skus, expand, options }, listWithPropertiesOperationSpec);
    }
    /**
     * ListWithPropertiesNext
     * @param location The name of a supported Azure region.
     * @param publisherName A valid image publisher.
     * @param offer A valid image publisher offer.
     * @param skus A valid image SKU.
     * @param nextLink The nextLink from the previous successful call to the ListWithProperties method.
     * @param options The options parameters.
     */
    _listWithPropertiesNext(location, publisherName, offer, skus, nextLink, options) {
        return this.client.sendOperationRequest({ location, publisherName, offer, skus, nextLink, options }, listWithPropertiesNextOperationSpec);
    }
}
exports.VirtualMachineImagesImpl = VirtualMachineImagesImpl;
// Operation Specifications
const serializer = coreClient.createSerializer(Mappers, /* isXml */ false);
const getOperationSpec = {
    path: "/subscriptions/{subscriptionId}/providers/Microsoft.Compute/locations/{location}/publishers/{publisherName}/artifacttypes/vmimage/offers/{offer}/skus/{skus}/versions/{version}",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.VirtualMachineImage,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.location1,
        Parameters.publisherName,
        Parameters.offer,
        Parameters.skus,
        Parameters.version,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const listOperationSpec = {
    path: "/subscriptions/{subscriptionId}/providers/Microsoft.Compute/locations/{location}/publishers/{publisherName}/artifacttypes/vmimage/offers/{offer}/skus/{skus}/versions",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: {
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "VirtualMachineImageResource",
                        },
                    },
                },
            },
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    queryParameters: [
        Parameters.apiVersion,
        Parameters.expand1,
        Parameters.top,
        Parameters.orderby,
    ],
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.location1,
        Parameters.publisherName,
        Parameters.offer,
        Parameters.skus,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const listOffersOperationSpec = {
    path: "/subscriptions/{subscriptionId}/providers/Microsoft.Compute/locations/{location}/publishers/{publisherName}/artifacttypes/vmimage/offers",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: {
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "VirtualMachineImageResource",
                        },
                    },
                },
            },
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.location1,
        Parameters.publisherName,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const listPublishersOperationSpec = {
    path: "/subscriptions/{subscriptionId}/providers/Microsoft.Compute/locations/{location}/publishers",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: {
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "VirtualMachineImageResource",
                        },
                    },
                },
            },
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.location1,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const listSkusOperationSpec = {
    path: "/subscriptions/{subscriptionId}/providers/Microsoft.Compute/locations/{location}/publishers/{publisherName}/artifacttypes/vmimage/offers/{offer}/skus",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: {
                type: {
                    name: "Sequence",
                    element: {
                        type: {
                            name: "Composite",
                            className: "VirtualMachineImageResource",
                        },
                    },
                },
            },
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.location1,
        Parameters.publisherName,
        Parameters.offer,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const listByEdgeZoneOperationSpec = {
    path: "/subscriptions/{subscriptionId}/providers/Microsoft.Compute/locations/{location}/edgeZones/{edgeZone}/vmimages",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.VmImagesInEdgeZoneListResult,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    queryParameters: [Parameters.apiVersion],
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.location1,
        Parameters.edgeZone,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const listWithPropertiesOperationSpec = {
    path: "/subscriptions/{subscriptionId}/providers/Microsoft.Compute/locations/{location}/publishers/{publisherName}/artifacttypes/vmimage/offers/{offer}/skus/{skus}/versions",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.VirtualMachineImagesWithPropertiesListResult,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    queryParameters: [
        Parameters.apiVersion,
        Parameters.top,
        Parameters.orderby,
        Parameters.expand5,
    ],
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.location1,
        Parameters.publisherName,
        Parameters.offer,
        Parameters.skus,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
const listWithPropertiesNextOperationSpec = {
    path: "{nextLink}",
    httpMethod: "GET",
    responses: {
        200: {
            bodyMapper: Mappers.VirtualMachineImagesWithPropertiesListResult,
        },
        default: {
            bodyMapper: Mappers.CloudError,
        },
    },
    urlParameters: [
        Parameters.$host,
        Parameters.subscriptionId,
        Parameters.nextLink,
        Parameters.location1,
        Parameters.publisherName,
        Parameters.offer,
        Parameters.skus,
    ],
    headerParameters: [Parameters.accept],
    serializer,
};
